### PR TITLE
enable downstream libraries to turn off direct cudaq-mlir-runtime linking with nvq++

### PR DESCRIPTION
**Pull request by amccaskey**
When compiling quantum device code for libraries, it is beneficial for one to be able to avoid direct linking to libcudaq-mlir-runtime and libcudaq-builder. If removing these links is not possible, and one tries to expose compiled device code to Python, they will hit the LLVM static initialization errors due to LLVM/MLIR being provided twice (once in the python cudaq module, and then again from the device library). 

This PR enables one to turn off those links with a compiler flag. This PR also fixes one of the exp_pauli signatures (add const to the pauli_word argument) to fix downstream compilation issues.

### CODE REVIEW

**Comment by bmhowe23** [comment]
> When compiling quantum device code for libraries, it is beneficial for one to be able to avoid direct linking to libcudaq-mlir-runtime and libcudaq-builder.

Does that mean you won't be able to execute the code then? I'm trying to understand if this is for a corner case or if it is the "regular" library use case for you.

**Comment by amccaskey** [comment]
> > When compiling quantum device code for libraries, it is beneficial for one to be able to avoid direct linking to libcudaq-mlir-runtime and libcudaq-builder.
> 
> Does that mean you won't be able to execute the code then? I'm trying to understand if this is for a corner case or if it is the "regular" library use case for you.

This is a regular use case for enabling library re-use. Specifically if we build quantum device libraries in C++ and want to reuse that code at the Python level, then we will run into issues if we don't have manual control over linking the MLIR-related libraries. Let's say you want to create the following library: 
```cpp
// lib.h 
void my_library_kernel(cudaq::qubit& q, double param); 
// lib.cpp 
__qpu__ void my_library_kernel(cudaq::qubit& q, double param) { ry(param,q); }
```
and compile to a `libmy_device_lib.so` using `nvq++`. Without this PR, you will always link to `libcudaq-mlir-runtime` and `libcudaq-builder`. If you then decide to use this kernel library from your own custom Python bindings (so in the presence of an already imported `cudaq` python module), you will hit issues with LLVM static initialization (_pycudaq provides static initializers from LLVM and now so does your new Python module, this was the same reason for all the hacky `BaseRemote...` headers in cudaq-common).

As for executing code, I might not fully understand what you mean, but you should always still be able to execute the code.  You can compile individual kernel files to a library, and then when compiling an application against this library, nvq++ will link the correct MLIR-related libraries so that you can run on any of our supported targets. This PR is primarily about providing control to ensure the link process succeeds for libraries that want to contribute to Python from C++. But maybe I'm missing something here that you see and I don't?

**Comment by bmhowe23** [comment]
> As for executing code, I might not fully understand what you mean, but you should always still be able to execute the code. You can compile individual kernel files to a library, and then when compiling an application against this library, nvq++ will link the correct MLIR-related libraries so that you can run on any of our supported targets. This PR is primarily about providing control to ensure the link process succeeds for libraries that want to contribute to Python from C++. But maybe I'm missing something here that you see and I don't?

OK, that all makes sense...thanks! It sounds like this is required to make a C++ library usable by both languages, and the top-level application is responsible for linking in the appropriate runtime, and only at *that* point will it become runnable (i.e. able to "executing the code". That makes sense.

File: tools/nvqpp/nvq++.in [hunk 1]
```diff
@@ -448,6 +452,9 @@ while [ $# -ne 0 ]; do
 	--emulate | -emulate)
 		CUDAQ_EMULATE_REMOTE=true
 		;;
+	--disable-mlir-links | -disable-mlir-links)
+	    ENABLE_MLIR_LIB_LINKING=false 
+		;;
```
Comment (bmhowe23) [inline tools/nvqpp/nvq++.in:457]: Could you please add this option to the `show_help` function near line 150, too?

**Review by bmhowe23 — APPROVED** [review]
LGTM. Just one small request below...

File: tools/nvqpp/nvq++.in (hunk 1 above)
Comment (amccaskey) [reply tools/nvqpp/nvq++.in:457]: I think I did, am I missing another spot? Line 158 there I put this 
```bash 
--disable-mlir-links
    Disable linking to cudaq-mlir-runtime and cudaq-builder. 
```

Comment (bmhowe23) [reply tools/nvqpp/nvq++.in:457]: Oops, you're right! Sorry about that! Too much task switching :)

**Comment by schweitzpgi** [comment]
Would this problem go away if we used shared libraries?

**Comment by amccaskey** [comment]
> Would this problem go away if we used shared libraries?

I think so. But I think the issue there is that LLVM best practice is to use static libraries since dynamic linking had been shown to hurt performance. I think this problem could also be solved by a future runtime refactor that replaced all the libraries with a single libcudaq that links to LLVM/MLIR statically.

**Pull request merged**